pkg/anchor/writer/writer.go

```diff
@@ -558,39 +565,108 @@ func (c *Writer) getWitnesses(refs []*operation.Reference) ([]string, error) {
+				finalAnchorOrigin, anchorOriginIPNSAddressWithScheme, ipnsAddressChain)
+
+			break
+		} else {
+			logger.Debugf("Resolved anchor origin IPNS address %s from %s, but this doesn't match the original "+
+				"anchor origin IPNS address (%s). %s will be resolved next if the max number of resolve levels "+
+				"hasn't been hit yet. Current IPNS address chain: %v.",
+				newAnchorOriginIPNSAddress, currentAnchorOriginIPNSAddressWithScheme,
+				anchorOriginIPNSAddressWithScheme, newAnchorOriginIPNSAddress, ipnsAddressChain)
+			currentAnchorOriginIPNSAddressWithScheme = newAnchorOriginIPNSAddress
+		}
+	}
+
+	if finalAnchorOrigin == "" {
+		return "", fmt.Errorf("max number of recursive resolves (%d) reached. IPNS address chain: %v",
+			maxIPNSResolveLevels, ipnsAddressChain)
+	}
 
+	return finalAnchorOrigin, nil
+}
+
+func (c *Writer) resolveWebFingerResponseFromIPNS(
+	anchorOriginIPNSAddress string) (discoveryrest.WebFingerResponse, error) {
 	respBytes, err := c.ipfsReader.Read(fmt.Sprintf("/ipns/%s/.well-known/webfinger",
-		anchorOriginSplitBySlashes[len(anchorOriginSplitBySlashes)-1]))
+		anchorOriginIPNSAddress))
 	if err != nil {
-		return "", fmt.Errorf("failed to resolve WebFinger response via IPNS: %w", err)
+		return discoveryrest.WebFingerResponse{}, fmt.Errorf("failed to read from IPFS: %w ", err)
 	}
 
 	var webFingerResp discoveryrest.WebFingerResponse
 
 	err = json.Unmarshal(respBytes, &webFingerResp)
 	if err != nil {
-		return "", fmt.Errorf("failed to unmarshal WebFinger response: %w", err)
+		return discoveryrest.WebFingerResponse{}, fmt.Errorf("failed to unmarshal WebFinger response: %w", err)
 	}
 
-	witnessProperty, exists := webFingerResp.Properties[discoveryrest.WitnessType]
-	if !exists {
-		return "", fmt.Errorf("witness property missing from the WebFinger response "+
-			"obtained by resolving %s", anchorOrigin)
+	return webFingerResp, nil
+}
+
+func (c *Writer) resolveAnchorOriginIPNSAddressFromOriginEndpoint(originEndpointForSuffix string) (string, error) {
+	resp, err := c.httpClient.Get(originEndpointForSuffix)
+	if err != nil {
+		return "", fmt.Errorf("failed to do GET on origin endpoint %s: %w", originEndpointForSuffix, err)
 	}
 
-	witnessPropertyString, ok := witnessProperty.(string)
-	if !ok {
-		return "", fmt.Errorf("failed to assert witness property from the WebFinger response "+
-			"obtained by resolving %s as a string", anchorOrigin)
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response body from origin endpoint %s: %w",
+			originEndpointForSuffix, err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("received status code %d from origin endpoint %s",
+			resp.StatusCode, originEndpointForSuffix)
 	}
 
-	logger.Debugf("Successfully resolved anchor origin %s from %s", witnessPropertyString, anchorOrigin)
+	err = resp.Body.Close()
```

Comment:
Response body needs to be closed regardless of whether status not OK. You should add a defer close right before ioutil.ReadAll.

Reply:
Oops, good catch. Fixed!